### PR TITLE
Add support for uniform/storage texel buffers

### PR DESCRIPTION
**Pull request by asuonpaa**
Fixes https://github.com/google/amber/issues/861

### CODE REVIEW

File: src/vulkan/buffer_descriptor.cc [hunk 1]
```diff
@@ -47,12 +47,23 @@ Result BufferDescriptor::CreateResourceIfNeeded() {
...
+  VkBufferUsageFlags flags =
+      VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT;
+  if (IsUniformBuffer()) {
+    flags |= VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT;
+  } else if (IsStorageBuffer()) {
+    flags |= VK_BUFFER_USAGE_STORAGE_BUFFER_BIT;
+  } else if (IsUniformTexelBuffer()) {
+    flags |= VK_BUFFER_USAGE_UNIFORM_TEXEL_BUFFER_BIT;
+  } else {
+    assert(IsStorageTexelBuffer());
```
Comment (dj2) [inline src/vulkan/buffer_descriptor.cc:62]: Instead of `assert` this should return a failed Result.

File: src/vulkan/transfer_buffer.cc [hunk 2]
```diff
@@ -19,10 +19,20 @@
...
+                               uint32_t size_in_bytes,
+                               Format* format)
+    : Resource(device, size_in_bytes) {
+  if (format)
+    format_ = device->GetVkFormat(*format);
+}
 
 TransferBuffer::~TransferBuffer() {
+  if (view_ != VK_NULL_HANDLE) {
+    device_->GetPtrs()->vkDestroyBufferView(device_->GetVkDevice(), view_,
```
Comment (dj2) [inline src/vulkan/transfer_buffer.cc:32]: I don't think you need the `if()` check. From the spec:

```
If bufferView is not VK_NULL_HANDLE, bufferView must be a valid VkBufferView handle
```

which sounds like you can just pass in the VK_NULL_HANDLE

File: src/vulkan/transfer_buffer.h [hunk 3]
```diff
@@ -30,10 +31,13 @@ class Device;
 /// Wrapper around a Vulkan VkBuffer object.
 class TransferBuffer : public Resource {
  public:
-  TransferBuffer(Device* device, uint32_t size_in_bytes);
+  TransferBuffer(Device* device,
+                 uint32_t size_in_bytes,
+                 Format* format = nullptr);
```
Comment (dj2) [inline src/vulkan/transfer_buffer.h:36]: nit: Remove the default value and pass `nullptr` at the various callers. I find default params makes refactoring a lot harder.

File: docs/amber_script.md [hunk 4]
```diff
@@ -492,17 +494,17 @@ contain image attachment content, depth/stencil content, uniform buffers, etc.
   # Attach |buffer_name| as a storage image. The MIP level will have a base
   # value of |level|.
   BIND BUFFER {buffer_name} AS storage_image \
-      [ BASE_MIP_LEVEL _level_ (default 0) ]
+      DESCRIPTOR_SET _id_ BINDING _id_ [ BASE_MIP_LEVEL _level_ (default 0) ]
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply docs/amber_script.md:497]: It means the documentation was broken.

File: src/vulkan/buffer_descriptor.cc (hunk 1 above)
Comment (asuonpaa) [reply src/vulkan/buffer_descriptor.cc:62]: done

File: src/vulkan/transfer_buffer.cc (hunk 2 above)
Comment (asuonpaa) [reply src/vulkan/transfer_buffer.cc:32]: Same applies to vkDestroyBuffer so I've removed the null check from there too.

File: src/vulkan/transfer_buffer.h (hunk 3 above)
Comment (asuonpaa) [reply src/vulkan/transfer_buffer.h:36]: done

**Pull request merged**